[WIP] CountTable now returns 0 instead of 'key not found' for immutable get…

BontaVlad opened the pull request:
Fixes #10065. 
~~I left the mutable get requests raise Key Error because I did not know what the proper response would be there.~~
Now is possible to do
```
  var tbl = toCountTable("abracadabra")
  doAssert tbl['z'] == 0
  tbl['z'] = 1
  doAssert tbl['z'] == 1
```

krux02 commented on lib/pure/collections/tables.nim:
```diff
@@ -960,16 +963,14 @@ proc mget*[A](t: var CountTable[A], key: A): var int {.deprecated.} =
 
 proc getOrDefault*[A](t: CountTable[A], key: A): int =
-  ## retrieves the value at ``t[key]`` iff ``key`` is in ``t``. Otherwise, 0 (the
+  ## retrieves the value at ``t[key]`` if ``key`` is in ``t``. Otherwise, 0 (the
```
Any particular reason you change `iff` to `if`? "iff" reads as "if and only if".

BontaVlad replied:
no reason at all, will change it back.

GULPF replied:
Looks like you accidentally changed it back to `if` in the last commit, both here and below

GULPF commented:
> Asking questions and at the same time modifying the data I think that hides a lot of magic that as a user I might not expect.

Semantically it doesn't modify any data, since with this PR each unknown key has an implicit value of 0 anyway.

But the biggest issue is that it's often not clear from the call site whether or not a variable is `var` or not, and with this PR `table["unknown key"]` would crash for `var` even if the result wont be modified. For example, if a `table` is changed from `let` to `var` for some other reason it could break the program. This type of difference in behavior for `var`/`let` should generally be avoided.

BontaVlad commented:
I completely agree with you about the different behavior for ```var/let```.

Araq commented:
I think this means  #10065 does not need to be fixed, it's good the way it is, people should use ``getOrDefault``. Alternatively the "[] returns var T" variant for CountTables could be deprecated.

krux02 commented:
@Araq there is a pragma already there: `proc `[]`*[A](t: var CountTable[A], key: A): var int {.deprecatedGet.}`. So I guess it is already deprecated?

I am not 100% sure if @GULPF pointed this out in his comment. But the problem is, there is a ``proc `[]`(t: var CountTable, key: A): var int``, and this can't be implemented with a default value. So I would vote to remove this variant, so that it can't be selected via overload resolution and instead provide a setter: ``proc `[]=`(t: var CountTable; key: A; value: int)``, so that ``table["unknown key"] = 123`` compiles. But code like ``table["some key"] += 13``, where a reference to the value is required, will break.

Araq commented:
That's not what ``deprecatedGet`` means. This pragma can be removed nowadays.

andreaferretti commented:
In my opinion it makes perfect sense to initialize the value to 0 if missing. Semantically, a missing value has a count of 0. I also don't see any issue with the var/let difference.

`proc []*[A](t: var CountTable[A], key: A): var int` is the preferred overload and will work for a `var`. If that var is changed to `let`, the other overload `proc []*[A](t: CountTable[A], key: A): int`, which will not allow modifications. What's not to like?

Araq commented:
Ok, andreaferretti's analysis seems to be solid.

Araq commented:
This PR breaks bootstrapping.

Araq commented on lib/pure/collections/tables.nim:
```diff
@@ -933,43 +933,74 @@ proc rawGet[A](t: CountTable[A], key: A): int =
+  newSeq(n, len(t.data) * growthFactor)
+  for i in countup(0, high(t.data)):
+    if t.data[i].val != 0: rawInsert(t, n, t.data[i].key, t.data[i].val)
+    swap(t.data, n)
```
Wrong indentation here.

andreaferretti commented:
I should also mention that @GULPF says

> with this PR table["unknown key"] would crash for var even if the result wont be modified

but to me it is not clear why - hence I may be missing something

GULPF commented:
@andreaferretti To clarify, I was arguing for the same thing that you're arguing for (see my first comment). The `var`/`let` comment was about the original PR which raised a `KeyError` for `var` but not for `let`.

krux02 commented:
@andreaferretti gread now you have a table that depending on whether it is a let or a var, reading from it might insert values into it. I am sure this won't ever cause any problems.

BontaVlad commented:
I hit a brick wall working on this PR.

I come from a python/javascript world, while I know about pointers, memory locations etc. these where not my daily concerns, hence my interest towards languages as Nim.

That's enough of context, comming back to the PR.

I don't like the way I implemented the PR for several reasons:
First off:

For every question regarding an unknown key you insert the key into the table:
```
  var t_mut = toCountTable("abc")
  doAssert t_mut.counter == 3
  doAssert t_mut['d'] == 0
  doAssert t_mut['e'] == 0
  doAssert t_mut.counter == 5
```
And I also introduced a bug because of this line ```while t.data[h].val != 0```
```
  var t_mut = toCountTable("abc")
  doAssert t_mut.counter == 3
  doAssert t_mut['d'] == 0
  doAssert t_mut['d'] == 0
  doAssert t_mut.counter == 5
```

At first I tried to implement the default stuff without inserting but I was hit by these errors:
```this expression has no address, use unsafeAddr``` or ```this variables exceeds it stack frame```, btw tell me if I should open a issue because ```https://nim-lang.org/docs/manual/var_t_return.html``` does not exist anymore.
I understand why these exceptions are thrown but sadly my nim-fu is not good enought to come up with a better solution that this PR crap.

andreaferretti commented:
@krux02 Yeah, I agree it is not nice. Semantically, it is fine, but one may end up filling a lot of keys with 0 unnecessarily, consuming more memory than needed. What do you suggest?

krux02 commented:
@andreaferretti look at the PR based on this PR ^. That is what I suggest.

andreaferretti commented:
See my comment there. Not that I have a good solution anyway

narimiran commented:
#10260 was merged instead
